controllers/argocd/server/role.go

```diff
@@ -0,0 +1,411 @@
+		}
+
+		roleRequest.ObjectMeta.Namespace = nsName
+
+		if nsName == sr.Instance.Namespace {
+			roleRequest.Rules = getPolicyRulesForArgoCDNamespace()
+		} else {
+			roleRequest.Rules = getPolicyRulesForManagedNamespace()
+		}
+
+		desiredRole, err := permissions.RequestRole(roleRequest)
+		if err != nil {
+			sr.Logger.Error(err, "reconcileManagedRoles: failed to request role", "name", desiredRole.Name, "namespace", desiredRole.Namespace)
+			sr.Logger.V(1).Info("reconcileManagedRoles: one or more mutations could not be applied")
+			reconciliationErrors = append(reconciliationErrors, err)
+			continue
+		}
+
+		// create default role only if custom role is not set
+		if getCustomRoleName() == "" {
+			// role doesn't exist in the namespace, create it
+			existingRole, err := permissions.GetRole(desiredRole.Name, desiredRole.Namespace, sr.Client)
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					sr.Logger.Error(err, "reconcileManagedRoles: failed to retrieve role", "name", desiredRole.Name, "namespace", desiredRole.Namespace)
+					reconciliationErrors = append(reconciliationErrors, err)
+					continue
+				}
+
+				// Only set ownerReferences for roles in same namespace as ArgoCD CR
+				if sr.Instance.Namespace == desiredRole.Namespace {
+					if err = controllerutil.SetControllerReference(sr.Instance, desiredRole, sr.Scheme); err != nil {
+						sr.Logger.Error(err, "reconcileManagedRoles: failed to set owner reference for role", "name", desiredRole.Name, "namespace", desiredRole.Namespace)
+						reconciliationErrors = append(reconciliationErrors, err)
+					}
+				}
+
+				if err = permissions.CreateRole(desiredRole, sr.Client); err != nil {
+					sr.Logger.Error(err, "reconcileManagedRoles: failed to create role", "name", desiredRole.Name, "namespace", desiredRole.Namespace)
+					reconciliationErrors = append(reconciliationErrors, err)
+					continue
+				}
+				sr.Logger.V(0).Info("reconcileManagedRoles: role created", "name", desiredRole.Name, "namespace", desiredRole.Namespace)
+				continue
+			}
+
+			// difference in existing & desired role, reset it
+			if !reflect.DeepEqual(existingRole.Rules, desiredRole.Rules) {
+				existingRole.Rules = desiredRole.Rules
+				if err = permissions.UpdateRole(existingRole, sr.Client); err != nil {
+					sr.Logger.Error(err, "reconcileManagedRoles: failed to update role", "name", existingRole.Name, "namespace", existingRole.Namespace)
+					reconciliationErrors = append(reconciliationErrors, err)
+					continue
+				}
+				sr.Logger.V(0).Info("reconcileManagedRoles: role updated", "name", existingRole.Name, "namespace", existingRole.Namespace)
+			}
+			
+			// role found, no changes detected
+			continue
+		} else {
+			// custom role in use, cleanup default role
+			err := sr.deleteRole(desiredRole.Name, desiredRole.Namespace)
+			if err != nil {
+				reconciliationErrors = append(reconciliationErrors, err)
+			}
+			continue
+		}
```

Comment:
I think I would prefer to invert this if else so that we check if `getCustomRole() != nil` then we delete the existing role and continue. That way the majority of the logic does not have to be at a nested level, which reduces the probability of bugs slightly and makes it more maintainable IMO

WDYT?

Reply:
Yes. Makes sense.